backend/pkg/modules/internal/bfla/bfladetector/learn_and_detect_bfla.go

```diff
@@ -404,26 +517,37 @@ func (l *learnAndDetectBFLA) traceRunner(ctx context.Context, trace *CompositeTr
 	if SpecTypeFromAPIInfo(apiInfo) == SpecTypeNone {
 		return fmt.Errorf("spec not present cannot learn BFLA; apiID=%d", trace.APIEvent.APIInfoID)
 	}
-	var tracesProcessed int
-	tracesProcessedEntry, err := l.tracesCounterMap.Get(apiID)
+
+	state, stateValue, err := l.checkBFLAState(apiID, []BFLAStateEnum{BFLALearning, BFLADetecting})
+
 	if err != nil {
-		log.Warnf("Could not load processed traces number: %s", err)
-	} else {
-		tracesProcessed, _ = tracesProcessedEntry.Get().(int)
+		return fmt.Errorf("unable to handle traces in the current state: %w", err)
 	}
-
-	if decrement, ok := l.mustLearn(apiID); ok {
-		log.Debugf("api %d; processed: %d", trace.APIEvent.APIInfoID, tracesProcessed)
-		// to still learn
+	if state.state == BFLALearning {
+		/* We are in the learning state */
+		log.Debugf("api %d; To process: %d", trace.APIEvent.APIInfoID, state.traceCounter)
 		err := l.updateAuthorizationModel(tags, resolvedPath, string(trace.APIEvent.Method),
 			trace.K8SSource, trace.APIEvent.APIInfoID, trace.DetectedUser, true, false)
 		if err != nil {
 			return err
 		}
 
-		decrement()
+		if state.traceCounter == -1 {
+			return nil
+		}
+		state.traceCounter--
```

Comment:
you should set the state to the new decremented value every time a trace is encountered, this should not have big performance issues since the persistent map will do updates only once in a while